### PR TITLE
data model: implement contribution

### PR DESCRIPTION
**Pull request by rerowep**
Implemnet: contribution

* Implements transformation for Marc21.
* Implements transformation for UNIMARC.
* Adapts transformation for eBooks.
* Adapts ES mapping.
* Adapts JSON schemas.
* Adapts detailed document view

Co-Authored-by: Peter Weber <peter.weber@rero.ch>

## Why are you opening this PR?

- Which task/US does it implement?
- Which issue does it fix?

## Dependencies

My PR depends on the following `rero-ils-ui`'s PR(s):

* rero/rero-ils-ui#<xx>

## How to test?

- What command should I have to run to test your PR?
- What should I test through the UI?

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?

### CODE REVIEW

File: rero_ils/modules/documents/jsonschemas/documents/document_contribution-v0.0.1.json [hunk 1]
```diff
@@ -0,0 +1,49 @@
+{
+  "contribution": {
+    "title": "Contribution",
+    "description": "Relationship to an agent associated with the document",
+    "type": "array",
+    "minItems": 1,
+    "items": {
+      "type": "object",
+      "title": "Agents",
+      "description": "Person, family or corporate body (including conferences). Always create a link to IdRef or GND, if possible.",
```
Comment (jma) [inline rero_ils/modules/documents/jsonschemas/documents/document_contribution-v0.0.1.json:10]: additionnalProperties is missing

File: rero_ils/modules/documents/jsonschemas/documents/document_contribution_identifiedby-v0.0.1.json [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  ],
+  "required": [
+    "value"
+  ],
+  "properties": {
+    "type": {
+      "title": "Type",
+      "type": "string",
+      "default": "bf:Local",
+      "enum": [
```
Comment (jma) [inline rero_ils/modules/documents/jsonschemas/documents/document_contribution_identifiedby-v0.0.1.json:18]: you can use const instead

File: rero_ils/modules/documents/jsonschemas/documents/document_contribution_identifiedby-v0.0.1.json [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+          }
+        ],
+        "templateOptions": {
+          "cssClass": "col-lg-4"
+        }
+      }
+    },
+    "source": {
+      "title": "Source",
+      "type": "string",
```
Comment (jma) [inline rero_ils/modules/documents/jsonschemas/documents/document_contribution_identifiedby-v0.0.1.json:35]: minLenght is missing

File: rero_ils/modules/documents/jsonschemas/documents/document_contribution_person-v0.0.1.json [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+      "form": {
+        "placeholder": "Example: M\u00fcller, Hans",
+        "templateOptions": {
+          "cssClass": "col-lg-6"
+        }
+      }
+    },
+    "date_of_birth": {
+      "title": "Birth date",
+      "description": "Record only years.",
```
Comment (jma) [inline rero_ils/modules/documents/jsonschemas/documents/document_contribution_person-v0.0.1.json:45]: seems a bad description

File: rero_ils/modules/documents/jsonschemas/documents/document_contribution_person-v0.0.1.json [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+      "type": "string",
+      "minLength": 1,
```
Comment (jma) [inline rero_ils/modules/documents/jsonschemas/documents/document_contribution_person-v0.0.1.json:47]: can be 4 right?

File: rero_ils/modules/documents/mappings/v6/documents/document-v0.0.1.json [hunk 6]
```diff
@@ -54,6 +54,17 @@
...
               }
             },
+            "partOfTitle": {
+              "properties": {
+                "value": {
+                  "type": "text",
+                  "index": false
+                },
+                "language": {
+                  "type": "keyword"
```
Comment (jma) [inline rero_ils/modules/documents/mappings/v6/documents/document-v0.0.1.json:64]: do we need to index this property?

File: rero_ils/modules/documents/utils.py [hunk 7]
```diff
@@ -398,3 +390,72 @@ def title_format_text(title, with_subtitle=True):
         else:
             title_text.append({'value': value, 'language': key})
     return title_text
+
+
+def create_authorized_access_point(agent):
+    """Create the authorized_access_point for an agent.
+
+    :param agent: Agent to create the authorized_access_point for.
```
Comment (jma) [inline rero_ils/modules/documents/utils.py:398]: :return is missing.

File: rero_ils/modules/persons/api.py [hunk 8]
```diff
@@ -222,6 +214,13 @@ def organisation_pids(self):
                 organisations.add(result.key)
         return list(organisations)
 
+    def get_authorized_access_point(self, language):
+        """Get localized authorized_access_point."""
```
Comment (jma) [inline rero_ils/modules/persons/api.py:218]: param and return is missing

File: rero_ils/utils.py [hunk 9]
```diff
@@ -29,3 +30,10 @@ def get_current_language():
     """Return the current selected locale."""
     loc = current_i18n.locale
     return loc.language
+
+
+def get_i18n_supported_languages():
+    """Get defined languages from config."""
```
Comment (jma) [inline rero_ils/utils.py:36]: return is  missing

File: rero_ils/modules/documents/jsonschemas/documents/document_contribution-v0.0.1.json (hunk 1 above)
Comment (rerowep) [reply rero_ils/modules/documents/jsonschemas/documents/document_contribution-v0.0.1.json:10]: ok

File: rero_ils/modules/documents/jsonschemas/documents/document_contribution_identifiedby-v0.0.1.json (hunk 2 above)
Comment (rerowep) [reply rero_ils/modules/documents/jsonschemas/documents/document_contribution_identifiedby-v0.0.1.json:18]: ok

File: rero_ils/modules/documents/jsonschemas/documents/document_contribution_person-v0.0.1.json (hunk 4 above)
Comment (rerowep) [reply rero_ils/modules/documents/jsonschemas/documents/document_contribution_person-v0.0.1.json:45]: It is in https://docs.google.com/spreadsheets/d/1DCjCcMMylVsSoKIjsjx6ykWnGr9r_wPZ0fLRi3rlVpQ/edit#gid=1498006006 like this.

File: rero_ils/modules/documents/jsonschemas/documents/document_contribution_identifiedby-v0.0.1.json (hunk 3 above)
Comment (rerowep) [reply rero_ils/modules/documents/jsonschemas/documents/document_contribution_identifiedby-v0.0.1.json:35]: ok

File: rero_ils/modules/documents/jsonschemas/documents/document_contribution_person-v0.0.1.json (hunk 5 above)
Comment (rerowep) [reply rero_ils/modules/documents/jsonschemas/documents/document_contribution_person-v0.0.1.json:47]: Not if you have books from the romans 400.

File: rero_ils/modules/documents/utils.py (hunk 7 above)
Comment (rerowep) [reply rero_ils/modules/documents/utils.py:398]: ok

File: rero_ils/modules/documents/mappings/v6/documents/document-v0.0.1.json (hunk 6 above)
Comment (rerowep) [reply rero_ils/modules/documents/mappings/v6/documents/document-v0.0.1.json:64]: ok

File: rero_ils/modules/persons/api.py (hunk 8 above)
Comment (rerowep) [reply rero_ils/modules/persons/api.py:218]: ok

File: rero_ils/utils.py (hunk 9 above)
Comment (rerowep) [reply rero_ils/utils.py:36]: ok

File: rero_ils/modules/patrons/templates/rero_ils/_patron_profile_loans.html [hunk 10]
```diff
@@ -90,10 +90,10 @@ <h6 class="mt-2 mb-2 ng-star-inserted">{{ _('Details') }}</h6>
         {% if loan.document.authors %}
```
Comment (Garfield-fr) [inline rero_ils/modules/patrons/templates/rero_ils/_patron_profile_loans.html:90]: Change with this `{% if loan.document.contribution %}`

File: rero_ils/modules/patrons/templates/rero_ils/_patron_profile_requests.html [hunk 11]
```diff
@@ -79,10 +79,10 @@ <h6 class="mt-2 mb-2 ng-star-inserted">{{ _('Details') }}</h6>
         {% if loan.document.authors %}
```
Comment (Garfield-fr) [inline rero_ils/modules/patrons/templates/rero_ils/_patron_profile_requests.html:79]: Change with this `{% if loan.document.contribution %}`

**Review by Garfield-fr — APPROVED** [review]

**Review by iGormilhit — CHANGES_REQUESTED** [review]
Could you mention https://github.com/rero/rero-ils/issues/1030 in the commit message, as I've been inform that your commit should fix it. Thanks.

**Review by pronguen — COMMENTED** [review]
* Field `role` should be required.
* When linking an agent to MEF, the source name should be displayed (it is not, see image below).
image
* When I edit an existing document ([this one for ex](https://ilsdev.test.rero.ch/professional/records/documents/detail/1212)) and add a MEF link with a role, I click on "Save" but it does not save. It displays the processing symbol indefinitely.
* The role should be displayed in the public detailed view (as it is in the pro detailed view)
* In some cases like place of a corporate body, parenthesis should be removed from the MARC field 710|711$c (see image below)
* The subordinate unit of corporate body should not be lost during conversion (compare https://ilsdev.test.rero.ch/api/documents/534 with http://data.rero.ch/01-R007226326/marchtml)(see image below)
* The message below the conference toggle button should not be displayed (see image below). It appears already as mouse hover.
image
* Corporate bodies as contributors are not displayed on pro detailed view. [Example](https://ilsdev.test.rero.ch/professional/records/documents/detail/796)
* In the pro and public brief view, the display of a corporate body with conference number
  * should not have a space after the parenthesis (see image below). This applies also to the public detailed view.
  * should not have the suspension points at the end (see image below)
  * should not be on a new line (see image below)
image
* It should be possible to add an identifier (`identifiedBy`) to a contributor (person or organisation).

**Review by iGormilhit — APPROVED** [review]

**Comment by rerowep** [comment]
> * conference toggle button should not be displayed (see image below). It appears already as mouse hover.

This is a known bug in `ng-core`.

**Comment by rerowep** [comment]
> * should not be on a new line (see image below)

The second line `Bruxelles ; New York : P.I.E. Peter Lang, 2010` is the provision activity. A separate line should be normal in this case?

**Comment by rerowep** [comment]
> * When linking an agent to MEF, the source name should be displayed (it is not, see image below).

The translations are missing, thats way it is not displayed. Will be solved with completed translations.

**Comment by rerowep** [comment]
> * It should be possible to add an identifier (`identifiedBy`) to a contributor (person or organisation).

It is implemented but we are unable to add hidden properties in a `oneOf` in the moment. We can always display `identifiedBy` or we find a way to implement it in `ng-core`

**Comment by pronguen** [comment]
* On the pro detailed view, a contribution of type person should be displayed with his/her birth and death date. It is not the case.
* Corporate bodies as contributors should be displayed on pro detailed view with their subordinate units (like in brief view). It is not the case.
* Corporate bodies as contributors should be displayed on pro detailed view with their number/place/date if available (like in brief view). It is not the case.
* In the pro and public brief view, the display of the contributions should not have the suspension points at the end when all contributions are shown (see image below)
image

**Pull request merged**